### DISCUSSION TITLE
A configuration question

### DISCUSSION THREAD

**Discussion by JTP-123** [facelessuser/pymdown-extensions#2192] [Q&A] [answered]
As a new beginner, I decided to make my small tiny flask blog support math formula, so I chose arithmatex under python-markdown.  After I have read <a href="https://facelessuser.github.io/pymdown-extensions/extensions/arithmatex/">offical documentation</a>, i made such configuration as follows, namely I put  mathjax cdn link at the bottom of the block script of base.html template, arithmatex is included by markdown function. But it doesn't work. Can anyone give me a favor? thank you.
![6](https://github.com/facelessuser/pymdown-extensions/assets/68136548/3789c7b7-2f3b-44c6-a963-12094203c7ff)
![7](https://github.com/facelessuser/pymdown-extensions/assets/68136548/e8e85675-3f45-4d89-bfbe-febb09729efb)
![5](https://github.com/facelessuser/pymdown-extensions/assets/68136548/c21d4f26-f8ff-4c4b-81ab-6fcf05fef893)

**Comment by facelessuser** [comment]
When providing examples, pictures are fine, but please always post code as copy/pastable code blocks.

As far as your question is concerned, why are you configuring `mdx_math` if you are using `pymdownx.arithmatex`?

**Comment by JTP-123** [comment]
I delete mdx_math configuration, the code now turns up as:
<code>
@staticmethod
def on_changed_body(target, value, oldvalue, initiator):
allowed_tags = ['a', 'abbr', 'acronym', 'b', 'blockquote', 'code', # define allowed html tags for Markdown
'em', 'i', 'li', 'ol', 'pre', 'strong', 'ul',
'h1', 'h2', 'h3', 'p', 'img', 'kbd', 'sup', 'sub', 'br',
'table', 'footer', 'thead', 'tbody', 'tr', 'th', 'td', 'small', 'big']
extensions = ['pymdownx.arithmatex', 'pymdownx.b64', 'pymdownx.emoji']
target.body_html = bleach.linkify(bleach.clean(
markdown(value, output_format='html',
extensions=extensions),
tags=allowed_tags, strip=True))
</code>
but it changes nothing.

**Reply by facelessuser** [reply, marked as answer]
Unfortunately, you've provided very little information that I can work with. You are clearly using a documentation system, one that I am not familiar with. I don't see your MathJax configuration anywhere, so it is likely you don't have it configured properly. Without you providing a minimal, reproducible example, I have no idea what you are doing.

With that said, I can provide a simple working example. This shows a simple, basic MathJax configuration and the setup required for Arithmatex. It will output an HTML file that you can load in your browser. You will see that MathJax works just fine with it.

```py
HTML = R"""
<!DOCTYPE html>
<html lang="en">

<body>
  %s
</body>

<script>
window.MathJax = {
  tex: {
    inlineMath: [ ["\\(","\\)"] ],
    displayMath: [ ["\\[","\\]"] ],
    processEscapes: true,
    processEnvironments: true
  },
  options: {
    ignoreHtmlClass: ".*",
    processHtmlClass: "arithmatex"
  }
};
</script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"></script>

</html>
"""

MD = R"""
$$
E(\mathbf{v}, \mathbf{h}) = -\sum_{i,j}w_{ij}v_i h_j - \sum_i b_i v_i - \sum_j c_j h_j
$$
"""

import markdown
result =  markdown.markdown(
    MD,
    extensions=['pymdownx.arithmatex'],
    extension_configs={
        'pymdownx.arithmatex': {
            'generic': True
        }
    }
)

with open('test.html', 'w', encoding='utf8') as f:
    f.write(HTML % result)
```

**Reply by JTP-123** [reply]
Indeed, it works fine with your own example code. I copied your configuration code into corresponding place at base.html  like this:
<code>
<script>
        window.MathJax = {
            tex: {
                inlineMath: [ ["\\(","\\)"] ],
                displayMath: [ ["\\[","\\]"] ],
                processEscapes: true,
                processEnvironments: true
            },
            options: {
                ignoreHtmlClass: ".*",
                processHtmlClass: "arithmatex"
            }
        };
    </script>
    <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
    <script id="MathJax-script" async 
            src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"></script>
<code>

And the rest part is added into markdown function:
<code>
def on_changed_body(target, value, oldvalue, initiator):
        allowed_tags = ['a', 'abbr', 'acronym', 'b', 'blockquote', 'code', # define allowed html tags for Markdown
                        'em', 'i', 'li', 'ol', 'pre', 'strong', 'ul', 
                        'h1', 'h2', 'h3', 'p', 'img', 'kbd', 'sup', 'sub', 'br', 
                        'table', 'footer', 'thead', 'tbody', 'tr', 'th', 'td', 'small', 'big']
        extensions = ['pymdownx.arithmatex', 'pymdownx.b64', 'pymdownx.emoji']
        target.body_html = bleach.linkify(bleach.clean(
            markdown(value, output_format='html', 
                    extensions=extensions,
                    extension_configs={
                        "pymdownx.arithmatex": {
                            "generic": True
                        }
                    }),
            tags=allowed_tags, strip=True))
</code>

I guess, like you said, Maybe the information provided is not enough to figure out why it does not work.

**Reply by facelessuser** [reply]
The issue is somewhere in your flask app. From my perspective, I've demonstrated that Arithmatex does work when configured properly. If you configure things the same way, it will work, but you could be stripping something important out with your app, sanitizing something that shouldn't be, or any number of things. I would look at your HTML output closely and see if it actually matches what you think it should. You can use the results from the script I gave you to compare against.

Unfortunately, my support can only extend as far as making sure it is clear how to use my extensions and ensuring they are working as expected. I cannot debug people's custom apps.

**Reply by JTP-123** [reply]
I finally solved this question, and thank you so much for your help.

**Comment by JTP-123** [comment]
But i still really appreciate your help, thank you so much.

**Reply by JTP-123** [reply]
I'm sorry for asking for help again. A quirk occurs. I mentioned that i have solved this question at last time. but this time when I input same math formula like <code>$p(x|y) = \frac{p(y|x)p(x)}{p(y)}$</code>,  It produces an unexpected  different output, that is <code><span class="arithmatex">\(p(x|y) = \frac{p(y|x)p(x)}{p(y)}\)</span></code>. I can not understand why such same thing will produce two different results at different time under the process of same code.
![批注 2023-10-13 200110](https://github.com/facelessuser/pymdown-extensions/assets/68136548/9c880572-1f3b-4a2f-9ee7-3aeab05c71ee)

**Reply by facelessuser** [reply]
I have no idea how you "fixed" your issue last time. You still have not provided me with a minimal, reproducible example. From what I understand, your entire app is custom. I have no context to help me understand what you are doing wrong. All I know is that if configured properly, everything works:

```py
HTML = R"""
<!DOCTYPE html>
<html lang="en">

<body>
  %s
</body>

<script>
window.MathJax = {
  tex: {
    inlineMath: [ ["\\(","\\)"] ],
    displayMath: [ ["\\[","\\]"] ],
    processEscapes: true,
    processEnvironments: true
  },
  options: {
    ignoreHtmlClass: ".*",
    processHtmlClass: "arithmatex"
  }
};
</script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"></script>

</html>
"""

MD = R"""
$p(x|y) = \frac{p(y|x)p(x)}{p(y)}$

$$
p(x|y) = \frac{p(y|x)p(x)}{p(y)}
$$

$$
E(\mathbf{v}, \mathbf{h}) = -\sum_{i,j}w_{ij}v_i h_j - \sum_i b_i v_i - \sum_j c_j h_j
$$
"""

import markdown
result =  markdown.markdown(
    MD,
    extensions=['pymdownx.arithmatex'],
    extension_configs={
        'pymdownx.arithmatex': {
            'generic': True
        }
    }
)

with open('test.html', 'w', encoding='utf8') as f:
    f.write(HTML % result)
```

<img width="1111" alt="Screenshot 2023-10-13 at 8 02 42 AM" src="https://github.com/facelessuser/pymdown-extensions/assets/1055125/9a0039aa-703d-438b-b483-58d151708ffe">

**Comment by JTP-123** [comment]
If I typed <code>$$
E(\mathbf{v}, \mathbf{h}) = -\sum_{i,j}w_{ij}v_i h_j - \sum_i b_i v_i - \sum_j c_j h_j
$$</code>, then the output is a plain text that is composed by a div tag with class='arithmatex"

**Comment by JTP-123** [comment]
The fact is that i changed nothing and the code was like before, during that time, i thought the reason lies in bleach.clean and then read the docmentation at <a href="https://bleach.readthedocs.io/en/latest/clean.html#bleach.clean">bleach.clean</a>. 
I did some experiment under powershell. Any html tags not in allowed tags list will be stripped or escaped such as: `bleach.clean("<div><span class='arithmatex'>x=1</span></div>", tags=['div'])` will output a string `"<div>&lt;span class='arithmatex'&gt;x=1&lt;/span&gt;</div>"` after the sanitizing of `bleach.clean`.
I deleted `bleach.clean` function, but nothing changes.  I happen to find that when i typed two math formula, the web was reloaded at once for some reason and after several minuates, the math result presented. But during this time, the web was immediately returned and the result was merrly mathjax formula which was wrapped by `<div>` or `<span>` html tag

**Comment by JTP-123** [comment]
A screenshot is maybe more clearer 
![批注 2023-10-14 142622](https://github.com/facelessuser/pymdown-extensions/assets/68136548/b984519b-2598-48e4-ac96-ec2518400463)

**Comment by JTP-123** [comment]
My app is basicly rests on flask development book and is customizedby myself in some way, i'm sorry for that i do not know what the minimal, reproducible example is for debugging

**Reply by facelessuser** [reply]
Yes, but:

1. I simply cannot debug your issue without a small, minimal, runnable snippet of code that reproduces your issues.
2. I've given you an example showing that if things are done properly, they work. This means there is no issue in `pymdown-extensions`. The issue lies with your other code.
3. I offer my free time to support issues with `pymdown-extensions`, but I do not have the free time to help everyone debug their own personal code, if I did, I'd have no free time. If `pymdown-extensions` is working as described, the problem lies somewhere in your flask app. It could be issues with any number of things you are doing. I do not offer support for `flask`,  `bleach`, or any other 3rd party software, just the software I've written or sometimes if I am involved with that community. If the problem is not in my code, I do not have the time to offer debugging help.

I mentioned before that the script I gave you will output valid HTML that works. I also mentioned that you should first identify what the difference between your failing HTML is vs the the working HTML my working script generates. You should be able to quickly see if HTML is getting altered along the way.

For instance, if my script generates the following and you are not getting that in the final HTML, then you have a problem where you are changing things.

```html
<div class="arithmatex">\[
p(x|y) = \frac{p(y|x)p(x)}{p(y)}
\]</div>
```

If I generate the script tag to configure MathJax in my script and it looks like the example below, then your script should generate it looking the same. If it does not, or it is missing, then you've done something wrong:

```html
<script>
window.MathJax = {
  tex: {
    inlineMath: [ ["\\(","\\)"] ],
    displayMath: [ ["\\[","\\]"] ],
    processEscapes: true,
    processEnvironments: true
  },
  options: {
    ignoreHtmlClass: ".*",
    processHtmlClass: "arithmatex"
  }
};
</script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"></script>
```

I understand you are new at this, but I've provided a script you can easily compare against. It works, so there are no issues on my end. If you narrow down the issue as a `bleach` issue, then you should direct any questions you have to `bleach`. If you have narrowed down the problem to a `flask` issue, you should ask any questions you have to `flask`.

**Reply by JTP-123** [reply]
Thank you for your reply, i really appreciate it and i will turn to another side for help, thank you so much.
